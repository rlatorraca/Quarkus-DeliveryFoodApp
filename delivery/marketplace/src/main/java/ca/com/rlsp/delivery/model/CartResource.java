package ca.com.rlsp.delivery.model;


import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;

import ca.com.rlsp.delivery.dto.DishDTO;
import ca.com.rlsp.delivery.dto.DishOrderDTO;
import ca.com.rlsp.delivery.dto.OrderDoneDTO;
import ca.com.rlsp.delivery.dto.RestaurantDTO;
import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;

@Path("cart")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class CartResource {

    private static final String CLIENT = "a";

    @Inject
    PgPool clientPgPool;

    @Inject
    @Channel("orders")
    Emitter<OrderDoneDTO> emitterOrder;

    @GET
    public Uni<List<DishCart>> buscarcarrinho() {
        return DishCart.findCarrinho(clientPgPool, CLIENT);
    }

    @POST
    @Path("/dish/{idDish}")
    public Uni<Long> adicionarPrato(@PathParam("idDish") Long idDish) {
        //poderia retornar o pedido atual
        DishCart pc = new DishCart();
        pc.client = CLIENT;
        pc.dish = idDish;
        return DishCart.save(clientPgPool, CLIENT, idDish);

    }

    @POST
    @Path("/ordering")
    public Uni<Boolean> finalizar() {
        OrderDoneDTO order = new OrderDoneDTO();
        String client = CLIENT;
        order.client = client;
        List<DishCart> pratoCarrinho = DishCart.findCarrinho(clientPgPool, client).await().indefinitely();
        //Utilizar mapstruts
        List<DishOrderDTO> dishes = pratoCarrinho.stream().map(pc -> from(pc)).collect(Collectors.toList());
        order.dishes = dishes;

        RestaurantDTO Restaurant = new RestaurantDTO();
        Restaurant.name = "Restaurant Name";
        order.restaurant = Restaurant;
        emitterOrder.send(order);
        return DishCart.delete(clientPgPool, client);
    }

    private DishOrderDTO from(DishCart pratoCarrinho) {
        DishDTO dto = Dish.findById(clientPgPool, pratoCarrinho.dish).await().indefinitely();        
        return new DishOrderDTO(dto.name, dto.description, dto.price);
    }

} 