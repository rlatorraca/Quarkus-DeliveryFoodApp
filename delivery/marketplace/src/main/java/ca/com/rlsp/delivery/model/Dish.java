package ca.com.rlsp.delivery.model;

import java.math.BigDecimal;
import java.util.stream.StreamSupport;

import ca.com.rlsp.delivery.dto.DishDTO;
import io.smallrye.mutiny.Multi;
import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Row;
import io.vertx.mutiny.sqlclient.RowSet;
import io.vertx.mutiny.sqlclient.Tuple;
public class Dish {

	public Long id;

    public String name;

	public String description;
   
	public BigDecimal price;
	
	 public static Multi<DishDTO> findAll(PgPool pgPool) {
	        Uni<RowSet<Row>> preparedQuery = pgPool.preparedQuery("select * from dish").execute();
	        //return preparedQuery.onItem().transformToMulti(rowSet -> Multi.createFrom().items(() -> {
	        //    return StreamSupport.stream(rowSet.spliterator(), false);
	        //})).onItem().transform(DishDTO::from);
	        return unitToMulti(preparedQuery);
	    }

	

	    public static Multi<DishDTO> findAll(PgPool client, Long idRestaurant) {
	        Uni<RowSet<Row>> preparedQuery = (Uni<RowSet<Row>>) client
	                .preparedQuery("SELECT * FROM dish where dish.restaurant_id = $1 ORDER BY name ASC").execute(
	                        Tuple.of(idRestaurant));
	        return unitToMulti(preparedQuery);
	    }

	    private static Multi<DishDTO> unitToMulti(Uni<RowSet<Row>> queryResult) {
	        return queryResult.onItem()
	                .transformToMulti(set -> Multi.createFrom().items(() -> {
	                    return StreamSupport.stream(set.spliterator(), false);
	                }))
	                .onItem().transform(DishDTO::from);
	    }

	    public static Uni<DishDTO> findById(PgPool client, Long id) {
	        return client.preparedQuery("SELECT * FROM dish WHERE id = $1").execute(Tuple.of(id))
	                .map(RowSet::iterator)
	                .map(iterator -> iterator.hasNext() ? DishDTO.from(iterator.next()) : null);
	    }
}
