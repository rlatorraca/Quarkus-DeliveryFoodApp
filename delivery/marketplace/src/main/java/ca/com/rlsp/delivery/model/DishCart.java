package ca.com.rlsp.delivery.model;

import java.util.ArrayList;
import java.util.List;

import io.smallrye.mutiny.Uni;
import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Row;
import io.vertx.mutiny.sqlclient.Tuple;

public class DishCart {

  public String client;

  public Long dish;
  
  public static Uni<Long> save(PgPool client, String cliente, Long prato) {
      return client.preparedQuery("INSERT INTO dish_client (client, dish) VALUES ($1, $2) RETURNING (client)").execute(
              Tuple.of(cliente, prato))

              .map(pgRowSet -> pgRowSet.iterator().next().getLong("cliente"));  
  }

  public static Uni<List<DishCart>> findCarrinho(PgPool client, String cliente) {
	  return client.preparedQuery("select * from dish_client where client = $1 ").execute(Tuple.of(cliente))
              .map(pgRowSet -> {
                  List<DishCart> list = new ArrayList<>(pgRowSet.size());
                  for (Row row : pgRowSet) {
                      list.add(toPratoCarrinho(row));
                  }
                  return list;
              });
  }

  private static DishCart toPratoCarrinho(Row row) {
	  DishCart pc = new DishCart();
      pc.client = row.getString("client");
      pc.dish = row.getLong("dish");
      return pc;
  }

  public static Uni<Boolean> delete(PgPool client, String cliente) {
	  return client.preparedQuery("DELETE FROM dish_client WHERE client = $1").execute(Tuple.of(cliente))
              .map(pgRowSet -> pgRowSet.rowCount() == 1);

  }
}
