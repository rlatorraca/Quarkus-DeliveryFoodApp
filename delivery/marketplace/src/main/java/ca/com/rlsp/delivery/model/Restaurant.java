package ca.com.rlsp.delivery.model;

import io.vertx.mutiny.pgclient.PgPool;
import io.vertx.mutiny.sqlclient.Tuple;

public class Restaurant {

	public Long id;

    public String name;

    public Mapposition mapPosition;
    
    @Override
    public String toString() {
        return "Restaurant [id=" + id + ", name=" + name + ", MapPosition=" + mapPosition+ "]";
    }
    
    public void persist(PgPool pgPool) {
        pgPool.preparedQuery("insert into mapposition (id, latitude, longetude) values ($1, $2, $3)")
        		.execute(Tuple.of(mapPosition.id, mapPosition.latitude, mapPosition.longetude))
        		.await().indefinitely();

        pgPool.preparedQuery("insert into restaurant (id, name, mapposition_id) values ($1, $2, $3)")
        		.execute(Tuple.of(id, name, mapPosition.id))
        		.await().indefinitely();

    }
}
