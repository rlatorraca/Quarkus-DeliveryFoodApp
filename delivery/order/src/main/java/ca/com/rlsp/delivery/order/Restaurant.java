package ca.com.rlsp.delivery.order;

import java.math.BigDecimal;

public class Restaurant {

	public Long id;
	
	public String name;
	
    public String description;

    public BigDecimal price;
}
