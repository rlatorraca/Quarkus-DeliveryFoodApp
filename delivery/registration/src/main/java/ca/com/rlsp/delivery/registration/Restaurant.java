package ca.com.rlsp.delivery.registration;

import java.util.Date;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToOne;
import javax.persistence.Table;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;

@Entity
@Table(name= "restaurant")
public class Restaurant extends PanacheEntityBase {

	
	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	public Long id;
	
	public String owner;
	
	public String name;
	
	public String registerNumber;
	
	@OneToOne
	public Localization localization;
	
	@CreationTimestamp
	public Date createdDate;
	
	@UpdateTimestamp
	public Date updatedDate;
	
	
}
