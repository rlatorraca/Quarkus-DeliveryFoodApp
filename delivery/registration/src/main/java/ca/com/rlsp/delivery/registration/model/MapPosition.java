package ca.com.rlsp.delivery.registration.model;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Table;

import io.quarkus.hibernate.orm.panache.PanacheEntityBase;

@Entity
@Table(name= "map_position")
public class MapPosition extends PanacheEntityBase {


	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	public Long id;
	
	public double latitude;
	
	public double longetude;
}
