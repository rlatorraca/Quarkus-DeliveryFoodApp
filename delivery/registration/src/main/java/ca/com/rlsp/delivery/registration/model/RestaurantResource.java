package ca.com.rlsp.delivery.registration.model;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.annotation.security.RolesAllowed;
import javax.inject.Inject;
import javax.json.bind.Jsonb;
import javax.json.bind.JsonbBuilder;
import javax.transaction.Transactional;
import javax.validation.Valid;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;

import org.eclipse.microprofile.jwt.Claim;
import org.eclipse.microprofile.jwt.Claims;
import org.eclipse.microprofile.jwt.JsonWebToken;
import org.eclipse.microprofile.metrics.annotation.Counted;
import org.eclipse.microprofile.metrics.annotation.SimplyTimed;
import org.eclipse.microprofile.metrics.annotation.Timed;
import org.eclipse.microprofile.openapi.annotations.enums.SecuritySchemeType;
import org.eclipse.microprofile.openapi.annotations.media.Content;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlow;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlows;
import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirement;
import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirements;
import org.eclipse.microprofile.openapi.annotations.security.SecurityScheme;
import org.eclipse.microprofile.openapi.annotations.tags.Tag;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;

import ca.com.rlsp.delivery.registration.dto.AddDishDTO;
import ca.com.rlsp.delivery.registration.dto.AddRestaurantDTO;
import ca.com.rlsp.delivery.registration.dto.DishDTO;
import ca.com.rlsp.delivery.registration.dto.RestaurantDTO;
import ca.com.rlsp.delivery.registration.dto.UpdateDishPriceDTO;
import ca.com.rlsp.delivery.registration.dto.UpdateRestaurantDTO;
import ca.com.rlsp.delivery.registration.mapper.DishMapper;
import ca.com.rlsp.delivery.registration.mapper.RestaurantMapper;
import ca.com.rlsp.delivery.registration.utils.ConstraintViolationResponse;
import io.quarkus.security.ForbiddenException;

@Path("/restaurants")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Tag(name = "restaurant")
@RolesAllowed("owner")
@SecurityScheme(securitySchemeName = "delivery-oauth", type = SecuritySchemeType.OAUTH2, flows = @OAuthFlows(password = @OAuthFlow(tokenUrl = "http://localhost:8180/auth/realms/delivery/protocol/openid-connect/token")))
//@SecurityRequirement(name="delivery-oauth")
@SecurityRequirements(value = { @SecurityRequirement(name = "delivery-oauth", scopes = {}) })
//@SecurityRequirements(value = {@SecurityRequirement(name = "delivery-oauth")})
public class RestaurantResource {

	@Inject
	@Channel("restaurants")
	Emitter<Restaurant> emitter;

	@Inject
	JsonWebToken jwt;

	@Inject
	@Claim(standard = Claims.sub)
	String sub;

	@Inject
	RestaurantMapper restaurantMapper;

	@Inject
	DishMapper dishMapper;

	/******************************************
	 * Restaurants
	 ******************************************/

	@GET
	@Counted(name = "Search amount for Restaurants", description = "Counter for Restaurants searching on API")
	@SimplyTimed(name = "name", description = "Computing the time spent to run a searching for Restaurants")
	@Timed(name = "All computing time for do a searching")
	public List<RestaurantDTO> getRestaurants() {
		Stream<Restaurant> restaurants = Restaurant.streamAll();
		return restaurants.map(r -> restaurantMapper.toRestaurantDTO(r)).collect(Collectors.toList());
	}

	@POST
	@Transactional
	@APIResponse(responseCode = "201", description = "Restaurant added Successfully")
	@APIResponse(responseCode = "400", content = @Content(schema = @Schema(allOf = ConstraintViolationResponse.class))) // Print
																														// violations/constraints/error
																														// in
																														// JSON
	public Response addRestaurant(@Valid AddRestaurantDTO dto) {
		Restaurant restaurant = restaurantMapper.toRestaurant(dto);
		restaurant.owner = sub;
		restaurant.persist();
//		Jsonb jsonb = JsonbBuilder.create();
//		String json = jsonb.toJson(restaurant);
//		emitter.send(json);
		emitter.send(restaurant);
		// emitter.send(restaurant);
		return Response.status(Status.CREATED).build();

	}

	@PUT
	@Path("{id}")
	@Transactional
	public void updateestaurant(@PathParam("id") Long id, UpdateRestaurantDTO dto) {
		Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(id);
		if (restaurantOptional.isEmpty()) {
			throw new NotFoundException();
		}

		Restaurant restaurant = restaurantOptional.get();
	    
		if (!restaurant.owner.equals(sub)) {
            throw new ForbiddenException();
        }

		// MapStruct: aqui passo a referencia para ser atualizada
		restaurantMapper.toRestaurant(dto, restaurant);

		restaurant.persist();

	}

	@DELETE
	@Path("{id}")
	@Transactional
	public void deleteRestaurant(@Valid @PathParam("id") Long id) {
		Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(id);
		if (!restaurantOptional.get().owner.equals(sub)) {
            throw new ForbiddenException();
        }

		restaurantOptional.ifPresentOrElse(Restaurant::delete, () -> {
			throw new NotFoundException();
		});

	}

	/******************************************
	 * Dishes
	 ******************************************/

	// Searching for dishes
	@GET
	@Path("{idRestaurant}/dishes")
	@Tag(name = "dish")
	public List<DishDTO> getDishes(@PathParam("idRestaurant") Long idResraurant) {
		Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(idResraurant);

		if (restaurantOptional.isEmpty()) {
			throw new NotFoundException("Restaurant does not exist");
		}

		Stream<Dish> dish = Dish.stream("restaurant", restaurantOptional.get());
		return dish.map(p -> dishMapper.toDTO(p)).collect(Collectors.toList());

	}

	@POST
	@Path("{idRestaurant/dishes")
	@Transactional
	@Tag(name = "dish")
	public Response addDish(@PathParam("idRestaurant") Long idRestaurant, AddDishDTO dto) {
		Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(idRestaurant);

		if (restaurantOptional.isEmpty()) {
			throw new NotFoundException("Restaurant does not exist");
		}

		// Using DTO, receiving detached entity passed to persist()
		Dish dish = new Dish();
		dish.name = dto.name;
		dish.description = dto.description;
		dish.price = dto.price;
		dish.restaurant = restaurantOptional.get();

		dish.persist();

		return Response.status(Status.CREATED).build();
	}

	@PUT
	@Path("{idRestaurant/dishes/{id}")
	@Transactional
	@Tag(name = "dish")
	public void updateDish(@PathParam("idRestaurant") Long idRestaurant, @PathParam("id") Long id,
			UpdateDishPriceDTO dto) {
		Optional<Restaurant> restaurantOptional = Restaurant.findByIdOptional(idRestaurant);

		if (restaurantOptional.isEmpty()) {
			throw new NotFoundException("Restaurant does not exist");
		}

		// This application has uniques IDs for dishes
		Optional<Dish> dishOptional = Dish.findByIdOptional(id);

		if (dishOptional.isEmpty()) {
			throw new NotFoundException("Dish does not exist");
		}

		Dish dish = dishOptional.get();
		dishMapper.toDish(dto, dish); // Just update the price

		dish.persist();
	}

	@DELETE
	@Path("{idRestaurant}/{dishes/{id}")
	@Tag(name = "dish")
	public void deleteDish(@PathParam("idRestaurant") Long idRestaurant, @PathParam("id") Long id) {

		Optional<Restaurant> RestaurantOptional = Restaurant.findByIdOptional(idRestaurant);
		if (RestaurantOptional.isEmpty()) {
			throw new NotFoundException("Restaurant does not exist");
		}

		Optional<Dish> pratoOp = Dish.findByIdOptional(id);

		pratoOp.ifPresentOrElse(Dish::delete, () -> {
			throw new NotFoundException();
		});
	}

}