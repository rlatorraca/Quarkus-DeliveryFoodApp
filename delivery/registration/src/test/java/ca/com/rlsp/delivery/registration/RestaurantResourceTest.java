package ca.com.rlsp.delivery.registration;

import javax.ws.rs.core.Response.Status;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.jupiter.api.Test;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;

import ca.com.rlsp.delivery.registration.model.Restaurant;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.specification.RequestSpecification;

@DBRider
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE)
@QuarkusTest
@QuarkusTestResource(RegistrationTestLifeCycleManager.class)
public class RestaurantResourceTest {

    private String token;

//    @BeforeEach
//    public void gereToken() throws Exception {
//        token = TokenUtils.generateTokenString("/JWTProprietarioClaims.json", null);
//    }

    @Test
    @DataSet("restaurants-scenario-1.yml")
    public void testBuscarRestaurantes() {
        String resultado = given()
                .when().get("/restaurants")
                .then()
                .statusCode(Status.OK.getStatusCode())
                .extract().asString();
        Approvals.verifyJson(resultado);
    }
    
    

    private RequestSpecification given() {
        return RestAssured.given().contentType(ContentType.JSON);
    }

    //Test PUT example

    @Test
    @DataSet("restaurants-scenario-1.yml")
    public void testAlterarRestaurante() {
        Restaurant dto = new Restaurant();
        dto.name = "newName";
        Long parameterValue = 123L;
        String resultado = given()
                .with().pathParam("id", parameterValue)
                .body(dto)
                .when().put("/restaurants/{id}")
                .then()
                .statusCode(Status.NO_CONTENT.getStatusCode())
                .extract().asString();
        Assert.assertEquals("", resultado);
        Restaurant findById = Restaurant.findById(parameterValue);

        //poderia testar todos os outros atribudos
        Assert.assertEquals(dto.name, findById.name);

    }

}