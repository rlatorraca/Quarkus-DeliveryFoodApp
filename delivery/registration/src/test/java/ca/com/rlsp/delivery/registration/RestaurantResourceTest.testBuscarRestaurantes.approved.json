[
  {
    "id": 123,
    "name": "New Restaurant",
    "owner": "Rodrigo Latorraca"
  },
  {
    "id": 1234,
    "name": "Old Restaurant",
    "owner": "Matias Mancuzo"
  }
]